### ISSUE TITLE
html size in Inches(not pixel) not fitting to PDF

### ISSUE THREAD

**Issue by RiddP** [wkhtmltopdf/wkhtmltopdf#2991]
Hello,

Our html is developed to converted to 8.5 x 11 inch PDF(Letter). So we designed all sizes in Inches( and not in pixel) so that we can early compare generated PDF inch by inch. 
e.g our header of PDF is as below. which has marigins in left and right 0.5 in.

>  <div style="width:8in;height:1.5in; background-color:purple;margin-left:0px;margin-right:05in">
>            Header.......................
>   </div>

When we generates PDF, above div looks much smaller like 6.5In x 1.2in. So what is the login to make it exact in PDF ? Its looks like elements renders in PDF are less 0.8 times smaller. We use --zoom 1.2 but this not helping much.

Somebody body please suggest how we can render exact inches elements in PDF same as html ?

Re

**Comment by pallymore** [comment]
did you set all the margins to 0?

**Comment by PhilterPaper** [comment]
Please close your duplicate entry (#2990).

**Comment by RiddP** [comment]
yes we set all margins to 0 but still issue has not fixed. We have used --zoom 1.2 so its bit good now. but issues with fonts, line spacing, font styles as well. 16px looks bit small and we have to use approx 22px instead of 16 px. so what is ratio of our html and what PDF renders we are not sure.

(There is also another issue of using Open Sans fonts.)

This is second time we scrud up with wkhtmltopdf. We hope somebody will give us solution very soon.

**Comment by pallymore** [comment]
well have you tried debug mode? 
it might be good to check your css to make sure there are no surprises.

**Comment by RiddP** [comment]
Hmm. This is good idea. We will check with debug mode.

**Comment by RiddP** [comment]
Thanks for your answer. Debug mode did not helped much. 
Can you simply answer, what is ratio of html to PDF renders ? e.f if i need 8px font, i have to specify 8 x 12. x1.2 = 11.52px. So why this is happening ? 

How can i simply specify class and property e.g open-sand 8px; SemiBold fonts ? Is there any known issue of integrating open sans fonts while PDF generates from html ?

**Comment by pallymore** [comment]
You should use points in your CSS so it will be converted automatically. 
e.g. `font-size: 8pt;`

If you can see Open-sans in pdf mode, you should be fine.
I'm not familiar with zooming in wkhtmltopdf - never tried it. 

also, i usually wrap my content in a div, then specify the size for that div. I can also apply some styles that only shows up in debug mode so the web page looks almost exactly like a PDF file being displayed in the browser.

```html
<body>
    <div class='page'>
        Content goes here
    </div>
</body>
```

```sass
html,
body {
  margin: 0;
  padding: 0;
  text-rendering: optimizeLegibility;
}

.page {
  position: relative;
  page-break-after: always;
  margin: 0;
  padding: 0;
  width: 8.5in;
  height: 11in;
  overflow: hidden;
}
```

in html mode, I can use Chrome developer tools to easily tell if anything is wrong.

**Comment by RiddP** [comment]
We have created one demo html. 
<html lang="en" xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta charset="utf-8" />
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,800,700,600,300' rel='stylesheet' type='text/css'>
    <title></title>
    <style>
      
        @font-face{
             font-family:'Open Sans semibold';
            src : url('C:/abc/semibold/OpenSans-Semibold.ttf') format('truetype');
        }

        body {
              margin: 0;
              padding: 0;
              text-rendering: optimizeLegibility;
            }

        .semibold h1{
            font-family:'Open Sans semibold' sans-serif;
           
            font-style:normal;
            font-size:12.8pt;
        }

    </style>
</head>
<body>
    <div class="semibold">

        <h1>Client First Name & Last Name <br />Allow for Two Rows</h1>
      

    </div>
</body>

We are getting wrong result in pdf. Fonts are not proper and font size also not proper.
[test1.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/313855/test1.pdf)

**Comment by pallymore** [comment]
first of all, can you try to install the font on the machine, and remove the `font-face` definition. 
simply refer to the font with its PostScript name (not full name) from the open sans website, it seems the postscript name for your font choice should be `OpenSans-Semibold` 
also there is an issue with fallback fonts, you should only use one font in `font-family`

so, install font locally, remove font-face (dont worry we have another way to add it back), and change your `.semibold h1`:

```css
font-family: 'OpenSans-Semibold';
```

also, I don't think it makes sense to have "half-points". no printer can do that, 1 point should be the minimum size of something a printer can print. Not sure why you need 12.8 points?

**Comment by pallymore** [comment]
oh by the way...not sure why you have google fonts in your head, but also re-defining the font-face to find the font file in your local directory.

**Comment by pallymore** [comment]
[gh_1.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/314120/gh_1.pdf)

here's the file I got. (generated with wkhtmltopdf 0.12.2.1 OS X 64bit build)

command:

```bash
$ wkhtmltopdf --page-size Letter -B 0 -L 0 -R 0 -T 0 input.html output.pdf
```

**Comment by RiddP** [comment]
Hi Thanks for your reply.
When you check this pdf file right click on text and select Edit Text , you can then check the actual fonts on right side. Try to write text on same page with same fonts and font size, you will get something else and very different from generated text.

**Comment by RiddP** [comment]
[test12.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/314341/test12.pdf)
look at this pdf. First fonts that we have created with html and second one we write in pdf with same font and size we get from html. Huge difference in both fonts as they are same.

**Comment by RiddP** [comment]
We want Open Sans Semibold 16px that you can see in pdf last fonts are that. When we generate pdf 19.50px is smaller then 16px

**Comment by pallymore** [comment]
You will want to think in `points` when building PDFs (or files that are meant to be printed) 
16px - under standard PPI settings, equals to `12pt`. 

sorry I don't have a pdf editor. Rendering a PDF from HTML is very different from editing a PDF file from the tool. Even you are using the same font (in HTML and in PDF editor), they might be rendered with different methods - it's really difficult to have consistent results. 
In your test12.pdf, it seems the first row is generated with a serif font - which means your font setup is probably incorrect (for wkhtmltopdf)   have you checked the files I posted? those are generated with wkhtmltopdf, and I'm positive they are using the same fonts. 

![image](https://cloud.githubusercontent.com/assets/2566530/16050098/50b6fc46-3228-11e6-82ee-48037c38dacb.png)

In your test12.pdf - there are two different Semibold OpenSans fonts. One of them is inserted with PDF editor I think.

**Comment by RiddP** [comment]
First ( Client First Name...) is generated from wkhtmltopdf.
Second (Client First Name...) Is what i added from PDF Add Text tool to compare both fonts but they are different. We want to generate same font like Second (Client First Name...) from html. 
We tried all and still not get exact font from wkhtmltopdf.

**Comment by pallymore** [comment]
the font looks the same to me ... only the 2nd one is a bit larger. 

again, the text you add with PDF add text tool, is rendered differently with different methods.

**Comment by NicoBarbaros** [comment]
did you tried to put dpi to 96?

**Comment by Yurko-Fedoriv** [comment]
I believe I have resolved this with `--disable-smart-shrinking` option

**Comment by jacob8000** [comment]
@Yurko-Fedoriv's reply solves the problem.
This issue can be closed.
